src/language/printer.js

```diff
@@ -14,101 +14,96 @@ import { visit } from './visitor';
-        var selectionSet = node.selectionSet;
-        return !name ? selectionSet :
-          join([op, join([name, defs]), directives, selectionSet], ' ');
-      },
-      VariableDefinition: node =>
-        join([node.variable + ': ' + node.type, node.defaultValue], ' = '),
-      SelectionSet: node =>
-        length(node.selections) === 0 ? null :
-        indent('{\n' + join(node.selections, '\n')) + '\n}',
-      Field: node =>
-        join([
-          join([
-            join([
-              node.alias,
-              node.name
-            ], ': '),
-            manyList('(', node.arguments, ', ', ')')
-          ]),
-          join(node.directives, ' '),
-          node.selectionSet
-        ], ' '),
-      Argument: node => node.name + ': ' + node.value,
-
-      // Fragments
-
-      FragmentSpread: node =>
-        join(['...' + node.name, join(node.directives, ' ')], ' '),
-      InlineFragment: node =>
-        join([
-          '... on',
-          node.typeCondition,
-          join(node.directives, ' '),
-          node.selectionSet
-        ], ' '),
-      FragmentDefinition: node =>
-        join([
-          'fragment',
-          node.name,
-          'on',
-          node.typeCondition,
-          join(node.directives, ' '),
-          node.selectionSet
-        ], ' '),
-
-
-      // Value
-
-      IntValue: node => node.value,
-      FloatValue: node => node.value,
-      StringValue: node => JSON.stringify(node.value),
-      BooleanValue: node => node.value ? 'true' : 'false',
-      EnumValue: node => node.value,
-      ListValue: node => '[' + join(node.values, ', ') + ']',
-      ObjectValue: node => '{' + join(node.fields, ', ') + '}',
-      ObjectField: node => node.name + ': ' + node.value,
-
-      // Directive
-
-      Directive: node =>
-        join(['@' + node.name, manyList('(', node.arguments, ', ', ')')]),
-
-      // Type
-
-      NamedType: node => node.name,
-      ListType: node => '[' + node.type + ']',
-      NonNullType: node => node.type + '!',
-    }
-  });
+  return visit(ast, { leave: printDocASTReducer });
 }
 
-function indent(maybeString) {
-  return maybeString && maybeString.replace(/\n/g, '\n  ');
+export var printDocASTReducer = {
+  Name: node => node.value,
+  Variable: node => '$' + node.name,
+
+  // Document
+
+  Document: node => join(node.definitions, '\n\n') + '\n',
+  OperationDefinition(node) {
+    var op = node.operation;
+    var name = node.name;
+    var defs = many('(', node.variableDefinitions, ', ', ')');
+    var directives = join(node.directives, ' ');
+    var selectionSet = node.selectionSet;
+    return !name ? selectionSet :
+      join([op, join([name, defs]), directives, selectionSet], ' ');
+  },
+  VariableDefinition: node =>
+    join([node.variable + ': ' + node.type, node.defaultValue], ' = '),
+  SelectionSet: node => block(node.selections),
+  Field: node =>
+    join([
+      join([node.alias, node.name], ': ') +
```

Comment:
the more obvious construction won't work here since alias is optional.

This pattern of making a list of nullable tokens and then joining is used through the printer